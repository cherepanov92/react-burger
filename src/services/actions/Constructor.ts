import { OrderedIngredient } from '../../utils/types';

export const ADD_INGREDIENT: 'ADD_INGREDIENT' = 'ADD_INGREDIENT';
export const REMOVE_INGREDIENT: 'REMOVE_INGREDIENT' = 'REMOVE_INGREDIENT';
export const MOVE_INGREDIENT: 'MOVE_INGREDIENT' = 'MOVE_INGREDIENT';

interface IAddIngredient {
    readonly type: typeof ADD_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IRemoveIngredient {
    readonly type: typeof REMOVE_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IMoveIngredient {
    readonly type: typeof MOVE_INGREDIENT;
    ingredient: OrderedIngredient;
    newIndex: number;
}

export type TIngredientActions = IAddIngredient | IRemoveIngredient | IMoveIngredient;