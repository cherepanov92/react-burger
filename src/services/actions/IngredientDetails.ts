import { IngredientType } from '../../utils/types';

export const ADD_INGREDIENT_DETAILS: 'ADD_INGREDIENT_DETAILS' = 'ADD_INGREDIENT_DETAILS';
export const REMOVE_INGREDIENT_DETAILS: 'REMOVE_INGREDIENT_DETAILS' = 'REMOVE_INGREDIENT_DETAILS';

interface IAddIngredientDetails {
    readonly type: typeof ADD_INGREDIENT_DETAILS;
    ingredient: IngredientType;
}

interface IRemoveIngredientDetails {
    readonly type: typeof REMOVE_INGREDIENT_DETAILS;
}

export type TIngredientDetailsActions = IAddIngredientDetails | IRemoveIngredientDetails;