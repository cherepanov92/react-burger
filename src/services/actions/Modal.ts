import { EnumModalType } from '../../utils/types';

export const APPEND_MODAL_TYPE: 'APPEND_MODAL_TYPE' = 'APPEND_MODAL_TYPE';
export const APPEND_ERROR_MODAL_TYPE: 'APPEND_ERROR_MODAL_TYPE' = 'APPEND_ERROR_MODAL_TYPE';
export const REMOVE_MODAL_TYPE: 'REMOVE_MODAL_TYPE' = 'REMOVE_MODAL_TYPE';

export interface IAppendModalType {
    readonly type: typeof APPEND_MODAL_TYPE;
    modalType: EnumModalType;
}

interface IAppendErrorModalType {
    readonly type: typeof APPEND_ERROR_MODAL_TYPE;
    status?: string | null;
    message: string | null;
}

interface IRemoveModalTypeE {
    readonly type: typeof REMOVE_MODAL_TYPE;
}

export type TModalActions = IAppendModalType | IRemoveModalTypeE | IAppendErrorModalType;