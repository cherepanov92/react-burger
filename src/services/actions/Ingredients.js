import { getIngredients } from "../../utils/api";
import { APPEND_ERROR_MODAL_TYPE } from "./Modal";

export const GET_INGREDIENTS_REQUEST = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED = 'GET_INGREDIENTS_FAILED';

export const getIngredientsData = () => {
    return function(dispatch) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
        .then(res => {
            if (res && res.success) {
                dispatch({
                    type: GET_INGREDIENTS_SUCCESS,
                    items: res.data
                });
            } else {
                dispatch({
                    type: GET_INGREDIENTS_FAILED
                });
            }
        })
        .catch(err => {
            dispatch({
                type: APPEND_ERROR_MODAL_TYPE,
                errorStatus: err.status
            })
        })
    };
}