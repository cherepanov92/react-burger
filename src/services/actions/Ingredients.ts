import { getIngredients } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { IError, IngredientType } from '../../utils/types';
import { Dispatch } from 'react';

export const GET_INGREDIENTS_REQUEST: 'GET_INGREDIENTS_REQUEST' = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS: 'GET_INGREDIENTS_SUCCESS' = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED: 'GET_INGREDIENTS_FAILED' = 'GET_INGREDIENTS_FAILED';

interface IGetIngredientsRequest {
    readonly type: typeof GET_INGREDIENTS_REQUEST;
}

interface IGetIngredientsSuccess {
    readonly type: typeof GET_INGREDIENTS_SUCCESS;
    items: IngredientType[];
}

interface IGetIngredientsFailed {
    readonly type: typeof GET_INGREDIENTS_FAILED;
}

export type TIngredientsActions = IGetIngredientsRequest | IGetIngredientsSuccess | IGetIngredientsFailed;

export const getIngredientsData = () => {
    return function (dispatch: Dispatch<TIngredientsActions | IError>) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_INGREDIENTS_SUCCESS,
                        items: res.data
                    });
                } else {
                    dispatch({
                        type: GET_INGREDIENTS_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
