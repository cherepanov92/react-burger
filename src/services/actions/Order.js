import { sendOrder } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE } from './Modal';

export const SENT_ORDER_REQUEST = 'SENT_ORDER_REQUEST';
export const SENT_ORDER_SUCCESS = 'SENT_ORDER_SUCCESS';
export const SENT_ORDER_FAILED = 'SENT_ORDER_FAILED';

export const sendOrderRequest = orderList => {
    return function (dispatch) {
        dispatch({
            type: SENT_ORDER_REQUEST
        });
        sendOrder(orderList)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: SENT_ORDER_SUCCESS,
                        name: res.name,
                        number: res.order.number
                    });
                    dispatch({ type: APPEND_MODAL_TYPE, modalType: 'order' });
                } else {
                    dispatch({
                        type: SENT_ORDER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
