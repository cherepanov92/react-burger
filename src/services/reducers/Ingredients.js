import { GET_INGREDIENTS_REQUEST, GET_INGREDIENTS_SUCCESS, GET_INGREDIENTS_FAILED } from '../actions/Ingredients';

export default function ingredientsReducer(state = {}, action) {
    switch (action.type) {
        case GET_INGREDIENTS_REQUEST: {
            return {
                ...state,
                ingredientsRequest: true,
                ingredientsFailed: false
            };
        }
        case GET_INGREDIENTS_SUCCESS: {
            return {
                ...state,
                ingredients: action.items,
                ingredientsRequest: false
            };
        }
        case GET_INGREDIENTS_FAILED: {
            return {
                ...state,
                ingredientsFailed: true,
                ingredientsRequest: false
            };
        }
        default: {
            return state;
        }
    }
};
