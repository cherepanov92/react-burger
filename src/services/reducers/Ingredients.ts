import { GET_INGREDIENTS_REQUEST, GET_INGREDIENTS_SUCCESS, GET_INGREDIENTS_FAILED } from '../actions/Ingredients';
import { ingredientType } from '../../utils/types';

type ingredientsState = {
    ingredientsRequest?: false;
    ingredientsFailed?: false;
    ingredients?: ingredientType[];
};

export default function ingredientsReducer(state: ingredientsState = {}, action: any) {
    switch (action.type) {
        case GET_INGREDIENTS_REQUEST: {
            return {
                ...state,
                ingredientsRequest: true,
                ingredientsFailed: false
            };
        }
        case GET_INGREDIENTS_SUCCESS: {
            return {
                ...state,
                ingredients: action.items,
                ingredientsRequest: false
            };
        }
        case GET_INGREDIENTS_FAILED: {
            return {
                ...state,
                ingredientsFailed: true,
                ingredientsRequest: false
            };
        }
        default: {
            return state;
        }
    }
};
