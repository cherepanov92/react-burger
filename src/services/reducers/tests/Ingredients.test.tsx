import React from 'react';
import renderer from 'react-test-renderer';
import ingredientsReducer, { InitialIngredientsState as initialState } from '../Ingredients';
import { GET_INGREDIENTS_FAILED, GET_INGREDIENTS_REQUEST, GET_INGREDIENTS_SUCCESS } from '../../actions/Ingredients';
import { EnumIngredientType, IngredientType } from '../../../utils/types';

const ingredientItem: IngredientType = {
    _id: '60d3b41abdacab0026a733c6',
    name: 'Краторная булка N-200i',
    type: EnumIngredientType.BUN,
    proteins: 80,
    fat: 24,
    carbohydrates: 53,
    calories: 420,
    price: 1255,
    image: 'https://code.s3.yandex.net/react/code/bun-02.png',
    image_mobile: 'https://code.s3.yandex.net/react/code/bun-02-mobile.png',
    image_large: 'https://code.s3.yandex.net/react/code/bun-02-large.png',
    __v: 0
};

describe('IngredientsReducer:', () => {
    it('Начальное состояние', () => {
        expect(ingredientsReducer(initialState, {} as any)).toEqual(initialState);
        const tree = renderer.create(<>Начальное состояние</>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Запрос на получение списка ингредиентов', () => {
        expect(
            ingredientsReducer(initialState, {
                type: GET_INGREDIENTS_REQUEST
            })
        ).toEqual({
            ...initialState,
            ingredientsRequest: true,
            ingredientsFailed: false
        });
    });

    it('Список ингредиентов получен', () => {
        expect(
            ingredientsReducer(initialState, {
                type: GET_INGREDIENTS_SUCCESS,
                items: [ingredientItem]
            })
        ).toEqual({
            ...initialState,
            ingredients: [ingredientItem],
            ingredientsRequest: false
        });
    });

    it('Ошибка при получении списка ингредиентов', () => {
        expect(
            ingredientsReducer(initialState, {
                type: GET_INGREDIENTS_FAILED
            })
        ).toEqual({
            ...initialState,
            ingredientsFailed: true,
            ingredientsRequest: false
        });
    });
});
