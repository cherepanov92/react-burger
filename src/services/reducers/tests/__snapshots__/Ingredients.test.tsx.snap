// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientsReducer: Начальное состояние 1`] = `"Начальное состояние"`;
